client/discovery/discovery.go

```
@@ -0,0 +1,133 @@
+}
+
+// Context is used to initialize a discovery backend.
+type Context struct {
+	config *config.Config
+	logger *log.Logger
+	node   *structs.Node
+}
+
+// DiscoveryLayer wraps a set of discovery providers and provides
+// easy calls to register/deregister into all of them.
+type DiscoveryLayer struct {
+	Providers []Provider
+	Context
+}
+
+// NewDiscoveryLayer is used to initialize the discovery layer. It
+// automatically initializes all of the built-in providers and
+// checks their configuration.
+func NewDiscoveryLayer(
+	factories []Factory,
+	config *config.Config,
+	logger *log.Logger,
+	node *structs.Node) (*DiscoveryLayer, error) {
+
+	// Make the context
+	ctx := Context{
+		config: config,
+		logger: logger,
+		node:   node,
+	}
+
+	// Initialize the providers
+	var providers []Provider
+	for _, factory := range factories {
+		provider, err := factory(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed initializing %s discovery: %s",
+				provider.Name(), err)
+		}
+		if provider.Enabled() {
+			providers = append(providers, provider)
+		}
+	}
+	return &DiscoveryLayer{providers, ctx}, nil
+}
+
+// EnabledProviders returns the names of the enabled discovery providers.
+func (d *DiscoveryLayer) EnabledProviders() []string {
+	avail := make([]string, len(d.Providers))
+	for i, provider := range d.Providers {
+		avail[i] = provider.Name()
+	}
+	return avail
+}
+
+// Register iterates all of the providers and registers the given service
+// information with them. If an error is encountered, it is only logged to
+// prevent a single failure from crippling the entire discovery layer.
+func (d *DiscoveryLayer) Register(allocID string, parts []string, port int) {
+	for _, disc := range d.Providers {
+		name := disc.DiscoverName(parts)
+		if err := disc.Register(allocID, name, port); err != nil {
+			d.logger.Printf(
+				"[ERR] client.discovery: error registering %q with %s (alloc %s): %s",
+				parts, disc.Name(), allocID, err)
+			return
```

You probably don't want to return here because we potentially could register the service with other providers. It may be worth using multi error and storing all errors to return them. Latter on we could do a retry or some other logic rather than eating the error.